volrpynn/snn.py

```diff
@@ -0,0 +1,196 @@
+
+    refrac_count.masked_scatter_(spike.byte(), t_ref.masked_select(spike.byte()))
+    v.masked_scatter_(spike.byte(), v_reset.masked_select(spike.byte()))
+
+    return v, spike, refrac_count
+
+def adaptive_lif_step(v, beta, x, refrac_count, parameters, dt):
+    """Computes an euler integration update step of an adaptive lif
+    neuron equation.
+    """
+    v_th, v_rest, v_reset, t_ref, tau, tau_beta, delta_beta = parameters
+
+    v -= dt * tau * (v - v_rest)
+    beta -= dt * tau_beta * beta
+
+    v += torch.where(refrac_count == 0, x, torch.zeros_like(x))
+    refrac_count = torch.where(refrac_count > 0, refrac_count - dt, torch.zeros_like(refrac_count))
+    spike = heavy_side(v - (v_th + beta), torch.tensor([5.0])) # TODO: Figure out a good value for h
+
+    refrac_count.masked_scatter_(spike.byte(), t_ref.masked_select(spike.byte()))
+    v.masked_scatter_(spike.byte(), v_reset.masked_select(spike.byte()))
+    beta += delta_beta * spike
+
+    return v, spike, refrac_count, beta
+
+def lif_euler_integrate(x, weight, lif_parameters, num_timesteps, dt):
+    s = torch.zeros(num_timesteps, weight.shape[0])
+    v = torch.zeros(num_timesteps, weight.shape[0])
+    v[0] = lif_parameters.v_rest
+    refrac_count = torch.zeros(weight.shape[0])
+
+    lif_input = x.mm(weight.t())
+
+    for i in range(1,num_timesteps):
+        v[i], s[i], refrac_count = lif_step(v[i-1], lif_input[i], refrac_count, lif_parameters, dt)
+
+    return v,s
+
+class LIFTransferFunction(torch.autograd.function.Function):
+    """Implements a lif transfer function.
+
+    Note: Uses hand rolled euler integration, might be a better idea to use an existing
+    nest/genn implementation.
+    """
+
+    @staticmethod
+    def forward(ctx, input, weight, lif_parameters, num_timesteps, dt):
+        # integrate the lif equations for a given number of timesteps
+        voltages, spikes = lif_euler_integrate(input, weight, lif_parameters, num_timesteps, dt)
+        # TODO: Need to potentially save LIFParameters as a flat thing here,
+        # but we don't use them right now so we should be fine
+        ctx.save_for_backward(input, weight, voltages, spikes, torch.tensor(num_timesteps), torch.tensor(dt))
+        return spikes
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input, weight, voltages, spikes, lif_parameters, num_timesteps, dt = ctx.saved_tensors
+        grad_input = grad_weight = grad_voltages = None
+        grad_spikes = grad_lif_parameters = grad_num_timesteps = grad_dt = None
+
+
+        if ctx.needs_input_grad[0]:
+            grad_input = grad_output.mm(weight)
+        if ctx.needs_input_grad[1]:
+            grad_weight = grad_output.t().mm(input)
+
+        return grad_input, grad_weight, grad_voltages, grad_spikes, grad_lif_parameters, grad_num_timesteps, grad_dt
+
+lif_transfer_function = LIFTransferFunction.apply
+
+class LIF(torch.nn.Module):
```

Comment:
Cool concept. This could be the more general 'wrapper' class you talked about. And then we could abstract the implementation away.